backend/dataall/modules/datasets/services/dataset_service.py

```diff
@@ -50,7 +51,67 @@
 log = logging.getLogger(__name__)
 
 
+class DatasetServiceInterface(ABC):
+    @staticmethod
+    def check_before_delete(session, uri, **kwargs) -> bool:
+        """Abstract method to be implemented by dependent modules that want to add checks before deletion for dataset objects"""
+        return True
+
+    @staticmethod
+    def execute_on_delete(session, uri, **kwargs) -> bool:
+        """Abstract method to be implemented by dependent modules that want to add clean-up actions when a dataset object is deleted"""
+        return True
+
+    @staticmethod
+    def append_to_list_user_datasets(session, username, groups):
+        """Abstract method to be implemented by dependent modules that want to add datasets to the list_datasets that list all datasets that the user has access to"""
+        return True
+
+    @staticmethod
+    def resolve_additional_dataset_user_role(session, uri, username, groups):
+        """Abstract method to be implemented by dependent modules that want to add new types of user role in relation to a Dataset"""
+        return None
+
+    @staticmethod
+    def extend_attach_steward_permissions(session, dataset, new_stewards) -> bool:
+        """Abstract method to be implemented by dependent modules that want to attach additional permissions to Dataset stewards"""
+        return True
+
+    def extend_delete_steward_permissions(session, dataset, new_stewards) -> bool:
+        """Abstract method to be implemented by dependent modules that want to attach additional permissions to Dataset stewards"""
+        return True
+
+
 class DatasetService:
+    _interfaces: List[DatasetServiceInterface] = []
+
+    @classmethod
+    def register(cls, interface: DatasetServiceInterface):
+        cls._interfaces.append(interface)
+
+    @classmethod
+    def check_before_delete(cls, session, uri, **kwargs) -> bool:
+        """All actions from other modules that need to be executed before deletion"""
+        can_be_deleted = [interface.check_before_delete(session, uri, **kwargs) for interface in cls._interfaces]
+        return all(can_be_deleted)
+
+    @classmethod
+    def execute_on_delete(cls, session, uri, **kwargs) -> bool:
+        """All actions from other modules that need to be executed during deletion"""
+        for interface in cls._interfaces:
+            interface.execute_on_delete(session, uri, **kwargs)
+        return True
+
+    @classmethod
+    def _list_all_user_interface_datasets(cls, session, username, groups) -> List:
+        """All list_datasets from other modules that need to be appended to the list of datasets"""
+        all_subqueries = []
+        for interface in cls._interfaces:
+            interface_subquery = interface.append_to_list_user_datasets(session, username, groups)
+            if interface_subquery.first() is not None:
+                all_subqueries.append(interface_subquery)
+        return all_subqueries
```

Comment:
can this become...
```python
return [
        query
        for interface in cls._interfaces
        for query in interface.append_to_list_user_datasets(session, username, groups)
        if query is not None
    ]
```

Reply:
I think that loop is not correct because `interface.append_to_list_user_datasets(session, username, groups)` does not return a list of queries, it returns a single query. Also, the query itself won't be none, we need to check that it does not return any item. Maybe something like:


```
        return [
            interface.append_to_list_user_datasets(session, username, groups)
            for interface in cls._interfaces
            if interface.append_to_list_user_datasets(session, username, groups).first() is not None
        ]
```

Reply:
Shouldn't it be:
```python
return [
        query
        for interface in cls._interfaces
        for query in [interface.append_to_list_user_datasets(session, username, groups)]
        if query.first() is not None
    ]
```

Reply:
Then yes!